File: chatServer/lib/slugify.py
"""Deterministic name-to-slug converter for entity doc filenames.

``slugify("Sarah Chen")`` → ``"sarah-chen"``
``slugify("Acme Corp.")`` → ``"acme-corp"``
``slugify("O'Brien & Associates")`` → ``"obrien-associates"``

Unicode is NFKD-normalised and stripped to ASCII before kebab-casing.
The result is safe as a filename and doubles as a wikilink target.
"""

from __future__ import annotations

import re
import unicodedata

# Safety cap: filesystem limit is typically 255 bytes; leave margin.
_MAX_SLUG_LEN = 200


def slugify(name: str) -> str:
    """Convert a display name to a kebab-case filename slug."""
    if not name:
        return ""
    s = unicodedata.normalize("NFKD", name)
    s = s.encode("ascii", "ignore").decode("ascii")
    s = s.lower()
    s = re.sub(r"[^a-z0-9]+", "-", s)
    s = re.sub(r"-+", "-", s)
    s = s.strip("-")
    return s[:_MAX_SLUG_LEN]

File: webApp/src/lib/slugify.ts
/**
 * SPEC-053 AC-03: Deterministic name-to-slug converter for entity doc filenames.
 *
 * Mirrors the Python implementation in `chatServer/lib/slugify.py`.
 * Used for wikilink resolution and UI previews.
 *
 * "Sarah Chen" → "sarah-chen"
 * "Acme Corp." → "acme-corp"
 */

const MAX_SLUG_LEN = 200;

/**
 * Convert a display name to a kebab-case filename slug.
 *
 * Unicode is normalised to NFKD and non-ASCII characters are stripped.
 * Non-alphanumeric characters become hyphens; consecutive hyphens are
 * collapsed; leading/trailing hyphens are removed.
 */
export function slugify(name: string): string {
  if (!name) return '';
  // NFKD normalise and strip non-ASCII
  const normalised = name
    .normalize('NFKD')
    .split('')
    .filter((c) => c.charCodeAt(0) <= 127)
    .join('');
  return normalised
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '')
    .slice(0, MAX_SLUG_LEN);
}
